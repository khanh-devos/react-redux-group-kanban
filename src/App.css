@import url("https://fonts.googleapis.com/css?family=Roboto");

* {
  box-sizing: border-box;
  font-family: Roboto, fantasy;
  font-size: 14px;
  padding: 0;
  margin: 0.2rem;
}

:root {
  --bodyPadding: 15rem;

  @media screen and (max-width: 765px) {
    --bodyPadding: 1rem;
  }
}

body {
  text-align: center;
}

.App {
  padding: 1rem var(--bodyPadding);
}

.header {
  justify-content: space-between;
  flex-direction: row;
  display: flex;
  align-items: center;
}

.logo-container {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1rem;
}

.logo-img {
  width: 3rem;
}
