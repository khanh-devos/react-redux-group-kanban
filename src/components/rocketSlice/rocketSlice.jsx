import axios from 'axios';

const { createSlice, createAsyncThunk } = require('@reduxjs/toolkit');

const URL_ROCKET = 'https://api.spacexdata.com/v4/rockets';

const initialState = {
  isLoading: true,
  rockets: [],
  msg: null,
};

export const fetchGetRockets = createAsyncThunk(
  'rocket/fetchGetRockets',
  async (name, thunkApi) => {
    try {
      const res = await axios.get(URL_ROCKET);
      return res.data;
    } catch (err) {
      return thunkApi.rejectWithValue('all rockets not in the earth');
    }
  },
);

const rocketSlice = createSlice({
  name: 'rocket',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchGetRockets.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchGetRockets.fulfilled, (state, { payload }) => ({
        ...state,
        isLoading: false,
        rockets: payload.map(({
          name, id, type, flickr_images: flickrImages,
        }) => ({
          id, name, type, flickrImages,
        })),
      }))
      .addCase(fetchGetRockets.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

// export const {} = rocketSlice.actions;

export default rocketSlice.reducer;
