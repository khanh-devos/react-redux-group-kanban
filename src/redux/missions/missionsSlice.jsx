import axios from 'axios';

const { createSlice, createAsyncThunk } = require('@reduxjs/toolkit');

const URL_MISSIONS = 'https://api.spacexdata.com/v3/missions';

const initialState = {
  isLoading: true,
  missions: [],
  msg: null,
};

export const fetchMissions = createAsyncThunk(
  'missions/fetchMissions',
  async (name, thunkApi) => {
    try {
      const res = await axios.get(URL_MISSIONS);
      // console.log(res.data[0]);
      return res.data;
    } catch (err) {
      return thunkApi.rejectWithValue('all rockets not in the earth');
    }
  },
);

const missionsSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchMissions.fulfilled, (state, { payload }) => ({
        ...state,
        isLoading: false,
        missions: payload.map(({
          mission_id: missionId, mission_name: missionName, description,
        }) => ({
          missionId, missionName, description,
        })),
      }))
      .addCase(fetchMissions.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

export default missionsSlice.reducer;
