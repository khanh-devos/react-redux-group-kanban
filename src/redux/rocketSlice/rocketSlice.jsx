import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const URL_ROCKET = 'https://api.spacexdata.com/v4/rockets';

const initialState = {
  isLoading: true,
  rockets: [],
  msg: null,
};

export const fetchGetRockets = createAsyncThunk(
  'rocket/fetchGetRockets',
  async (name, thunkApi) => {
    try {
      const res = await axios.get(URL_ROCKET);
      return res.data;
    } catch (err) {
      return thunkApi.rejectWithValue('all rockets not in the earth');
    }
  },
);

const rocketSlice = createSlice({
  name: 'rocket',
  initialState,
  reducers: {
    toggleReserve: (state, { payload }) => {
      const newRockets = state.rockets.map((item) => {
        if (item.id === payload) return { ...item, reserved: !item.reserved || false };
        return item;
      });
      return {
        ...state,
        rockets: newRockets,
      };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchGetRockets.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchGetRockets.fulfilled, (state, { payload }) => ({
        ...state,
        isLoading: false,
        rockets: payload.map(({
          id, name, flickr_images: flickrImages, description,
        }) => ({
          id, name, flickrImages, description,
        })),
      }))
      .addCase(fetchGetRockets.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

export const { toggleReserve } = rocketSlice.actions;

export default rocketSlice.reducer;
