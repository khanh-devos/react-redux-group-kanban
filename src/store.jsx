import { configureStore } from '@reduxjs/toolkit';
import missionsReducer from './redux/missions/missionsSlice';
import rocketReducer from './redux/rocketSlice/rocketSlice';

const store = configureStore({
  reducer: {
    rocket: rocketReducer,
    mission: missionsReducer,
  },
});

export default store;
