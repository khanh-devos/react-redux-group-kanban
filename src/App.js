import './App.css';
import './components/missions/mission.css';
import './components/rockets/rocket.css';
import {
  BrowserRouter, Route, NavLink, Routes,
} from 'react-router-dom';
import Missions from './components/missions/Mission';
import Rockets from './components/rockets/Rocket';
import Logo from './planet.png';

function App() {
  const colorActiveRoute = ({ isActive }) => (isActive ? { textDecoration: 'underline' } : { textDecoration: 'none' });

  return (
    <div className="App">

      <BrowserRouter>
        <header className="header">
          <div className="logo-container">
            <img className="logo-img" src={Logo} alt="space logo" />
            <h1>Space Travelers &#39; Hub</h1>
          </div>

          <navigator>
            <NavLink style={colorActiveRoute} to="/">Rockets</NavLink>
            <NavLink style={colorActiveRoute} to="/mission">Missions</NavLink>
          </navigator>
        </header>

        <Routes>
          <Route path="/" element={<Rockets />} />
          <Route path="/mission" element={<Missions />} />

        </Routes>
      </BrowserRouter>
    </div>
  );
}

export default App;
