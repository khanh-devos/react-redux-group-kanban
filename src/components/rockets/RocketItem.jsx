import { PropTypes } from 'prop-types';
import { useDispatch } from 'react-redux';
import { toggleReserve } from '../../redux/rocketSlice/rocketSlice';

export default function RocketItem({
  image, name, description, reserved, id,
}) {
  const dispatch = useDispatch();
  const handleClick = (event) => {
    event.preventDefault();
    dispatch(toggleReserve(id));
  };

  return (
    <li className="rocket-li">
      <img src={image} className="rocket-li-imgs" alt="rocket" />
      <div className="rocket-li-name-desc">
        <h3 className="rocket-li-h3">{name}</h3>
        <p>
          {reserved && <b className="rocket-li-reserved-tag">reserved</b>}
          {description}
        </p>
        <button
          type="button"
          title="testClick"
          className={!reserved ? 'rocket-btn-type1' : 'rocket-btn-type2'}
          onClick={handleClick}
        >
          {!reserved ? 'Reserve Rocket' : 'Cancel Revervation' }
        </button>
      </div>
    </li>
  );
}

RocketItem.propTypes = {
  image: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
  id: PropTypes.string.isRequired,
};
