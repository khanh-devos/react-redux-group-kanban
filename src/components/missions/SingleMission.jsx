import { PropTypes } from 'prop-types';
import { useDispatch } from 'react-redux';
import { toggleJoinMission } from '../../redux/missions/missionsSlice';

export default function SingleMission({
  missionId, missionName, missionDescription, missionReserved,
}) {
  const dispatch = useDispatch();
  const handleJoinMission = (e) => {
    const missionId = e.target.id;
    dispatch(toggleJoinMission(missionId));
  };
  return (
    <tr>
      <td>
        {missionName}
      </td>
      <td>
        {missionDescription}
      </td>
      <td>
        { missionReserved && (
          <button
            type="button"
            className="mission-active-button"
          >
            Active Member
          </button>
        )}
        { !missionReserved && (
          <button
            type="button"
            className="mission-inactive-button"
          >
            NOT A MEMBER
          </button>
        )}
      </td>
      <td>
        { missionReserved && (
          <button
            id={missionId}
            type="button"
            onClick={handleJoinMission}
            className="leave-mission-button"
          >
            Leave Mission
          </button>
        )}
        { !missionReserved && (
          <button
            id={missionId}
            type="button"
            title="test-mission-btn"
            onClick={handleJoinMission}
            className="join-mission-button"
          >
            Join Mission
          </button>
        )}
      </td>
    </tr>
  );
}

SingleMission.propTypes = {
  missionId: PropTypes.string.isRequired,
  missionName: PropTypes.string.isRequired,
  missionDescription: PropTypes.string.isRequired,
  missionReserved: PropTypes.bool.isRequired,
};
