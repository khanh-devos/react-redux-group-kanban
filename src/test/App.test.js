import { fireEvent, render, screen } from '@testing-library/react';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import App from '../App';
import Missions from '../components/missions/Mission';
import SingleMission from '../components/missions/SingleMission';
import RocketItem from '../components/rockets/RocketItem';
import { URL_MISSIONS } from '../redux/missions/missionsSlice';
import { URL_ROCKET } from '../redux/rocketSlice/rocketSlice';
import store from '../store';

jest.mock('axios');

describe('test components', () => {
  const rockets = [
    {
      id: '111', name: 'rocket 1', flickr_images: ['http1'], description: 'desc1',
    },
    {
      id: '222', name: 'rocket 2', flickr_images: ['http2'], description: 'desc2',
    },
  ];
  const missions = [
    { mission_id: '111', mission_name: 'mission 1', description: 'desc1' },
    { mission_id: '222', mission_name: 'mission 2', description: 'desc2' },
  ];

  axios.get.mockImplementation((url) => {
    switch (url) {
      case URL_ROCKET:
        return Promise.resolve({ data: rockets });
      case URL_MISSIONS:
        return Promise.resolve({ data: missions });
      default:
        return Promise.reject(new Error('not found'));
    }
  });

  it('should match the snapshot', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <App />
        </Provider>,
      );
      expect(container).toMatchSnapshot();
      jest.clearAllMocks();
    });
  });

  
  it('should match Rocket', () => {
    const data = {
      image: 'http1', 
      name: 'rocket 1', 
      description: 'desc1', 
      reserved: false, 
      id: '111',
    }

    
    const { container } = render(
      <Provider store={store}>
        <RocketItem {...data} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
    
  })

  it('should match Mission', () => {
     const { container } = render(
      <Provider store={store}>
        <Missions />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  })

  
});
  


