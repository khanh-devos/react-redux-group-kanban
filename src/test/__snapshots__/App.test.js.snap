// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test components should match Mission 1`] = `
<div>
  <div
    class="mission-wrapper"
  >
    <table
      class="mission-table"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;

exports[`test components should match Rocket 1`] = `
<div>
  <li
    class="rocket-li"
  >
    <img
      alt="rocket"
      class="rocket-li-imgs"
      src="http1"
    />
    <div
      class="rocket-li-name-desc"
    >
      <h3
        class="rocket-li-h3"
      >
        rocket 1
      </h3>
      <p>
        desc1
      </p>
      <button
        class="rocket-btn-type1"
        title="testClick"
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </li>
</div>
`;

exports[`test components should match the snapshot 1`] = `<div />`;
